bodhi/tests/server/functional/test_updates.py

```diff
@@ -2520,6 +2520,353 @@ def test_edit_testing_update_reset_karma_with_same_tester(self, publish, *args):
+        self.db.flush()
+
+        up.comment(self.db, u'LGTM Now', author=u'ralph', karma=1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertFalse(up.has_negative_karma)
+
+        up.comment(self.db, u'wfm', author=u'luke', karma=1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertFalse(up.has_negative_karma)
+
+        up.comment(self.db, u'LGTM', author=u'puiterwijk', karma=1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertFalse(up.has_negative_karma)
+
+        self.assertEquals(up.karma, 3)
+        self.assertEquals(up.autokarma, False)
+
+        # Checks Push to Stable text in the html page for this update
+        id = 'kernel-3.11.5-300.fc17'
+        resp = self.app.get('/updates/%s' % id,
+                            headers={'Accept': 'text/html'})
+        self.assertIn('text/html', resp.headers['Content-Type'])
+        self.assertIn(id, resp)
+        self.assertIn('Push to Stable', resp)
+
+    @mock.patch(**mock_valid_requirements)
+    @mock.patch('bodhi.server.notifications.publish')
+    def test_autopush_non_critical_update_with_negative_karma(self, publish, *args):
+        """Autopush Non Critical update even though it receives negative karma"""
+        user = User(name=u'bob')
+        self.db.add(user)
+        self.db.flush()
+
+        nvr = u'bodhi-2.0.0-2.fc17'
+        args = self.get_update(nvr)
+        args['autokarma'] = True
+        args['stable_karma'] = 3
+        args['unstable_karma'] = -3
+
+        resp = self.app.post_json('/updates/', args)
+        self.assertEquals(resp.json['request'], 'testing')
+        publish.assert_called_with(topic='update.request.testing', msg=ANY)
+
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        up.status = UpdateStatus.testing
+        self.db.flush()
+
+        up.comment(self.db, u'Failed to work', author=u'ralph', karma=-1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertTrue(up.has_negative_karma)
+
+        up.comment(self.db, u'LGTM Now', author=u'ralph', karma=1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertFalse(up.has_negative_karma)
+
+        up.comment(self.db, u'wfm', author=u'luke', karma=1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertFalse(up.has_negative_karma)
+
+        up.comment(self.db, u'LGTM', author=u'puiterwijk', karma=1)
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertFalse(up.has_negative_karma)
+
+        self.assertEquals(up.karma, 3)
+        self.assertEquals(up.autokarma, True)
+
+        up = self.db.query(Update).filter_by(title=resp.json['title']).one()
+        self.assertEquals(up.request, UpdateRequest.stable)
```

Comment:
This test has the same user (ralph) giving -1 and then +1, which means that the -1 won't be counted as negative karma, right? Thus, this test actually doesn't have negative karma on the update it is testing.